tsl/src/hypercore/hypercore_handler.c

```diff
@@ -1787,6 +1789,87 @@ hypercore_tuple_delete(Relation relation, ItemPointer tid, CommandId cid, Snapsh
 	return result;
 }
 
+/*
+ * Decompress a segment that contains the row given by ctid.
+ *
+ * This function is typically called during an upsert (ON CONFLICT DO UPDATE),
+ * where the conflicting row points to a compressed segment that needs to be
+ * decompressed before the update can take place. This function is used to
+ * decompress that segment into a set of individual rows.
+ *
+ * Returns the number of rows in the segment that were decompressed, or 0 if
+ * the TID pointed to a regular (non-compressed) tuple. The TID of the
+ * de-compressed conflicting row is returned via "new_ctid".
+ */
+int
+hypercore_decompress_conflict_segment(Relation relation, const ItemPointer ctid,
+									  TupleTableSlot *slot, Snapshot snapshot, ItemPointer new_ctid)
+{
+	HypercoreInfo *hcinfo;
+	Relation crel;
+	TupleTableSlot *cslot;
+	ItemPointerData decoded_tid;
+	TM_Result result;
+	TM_FailureData tmfd;
+	int n_batch_rows = 0;
+	uint16 tuple_index;
+	bool should_free;
+
+	/* Nothing to do if this is not a compressed segment */
+	if (!is_compressed_tid(ctid))
+		return 0;
+
+	Assert(TTS_IS_ARROWTUPLE(slot));
+
+	hcinfo = RelationGetHypercoreInfo(relation);
+	crel = table_open(hcinfo->compressed_relid, RowExclusiveLock);
+	tuple_index = hypercore_tid_decode(&decoded_tid, ctid);
+
+	/* Is it guaranteed that "slot" is populated with the conflicting segment
+	 * tuple?  Otherwise we need to fetch the compressed segment... Unclear if
+	 * this is ever needed. */
+	if (TTS_EMPTY(slot) && !table_tuple_fetch_row_version(relation, ctid, snapshot, slot))
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("conflicting compressed segment not found")));
+
+	Assert(ItemPointerEquals(ctid, &slot->tts_tid));
+	cslot = arrow_slot_get_compressed_slot(slot, NULL);
+	HeapTuple tuple = ExecFetchSlotHeapTuple(cslot, false, &should_free);
+
+	RowDecompressor decompressor = build_decompressor(crel, relation);
```

Comment:
If it is possible to open the compressed table at the beginning of the statement, creating the decompressor could also be moved there. As before, not sure if there is a convenient place, so feel free to ignore.

Reply:
Unfortunately there is no "begin" for update and we aren't passed any scanstate that underlies the update here. We'd have to move the state outside this function, e.g., in hypertable_modify.c, but that moves more and more functionality outside the TAM, which we want to avoid, ideally.

Reply:
Thanks for checking.